controllers/dbaas.redhat.com/crunchybridgeinstance_controller.go

```diff
@@ -0,0 +1,311 @@
+//+kubebuilder:rbac:groups=dbaas.redhat.com,resources=crunchybridgeinstances/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=dbaas.redhat.com,resources=crunchybridgeinstances/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// the CrunchyBridgeInstance object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.9.2/pkg/reconcile
+func (r *CrunchyBridgeInstanceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	instanceObj := &dbaasredhatcomv1alpha1.CrunchyBridgeInstance{}
+	if err := r.Get(ctx, req.NamespacedName, instanceObj); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Likely deleted before action or extra pass post-deletion, no-op
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, "Error fetching CrunchyBridgeInstance object for reconciliation")
+		return ctrl.Result{}, err
+	}
+	inventory := dbaasredhatcomv1alpha1.CrunchyBridgeInventory{}
+	if err := r.Get(ctx, types.NamespacedName{Namespace: instanceObj.Spec.InventoryRef.Namespace, Name: instanceObj.Spec.InventoryRef.Name}, &inventory); err != nil {
+		if apierrors.IsNotFound(err) {
+			// CR deleted since request queued, child objects getting GC'd, no requeue
+			logger.Info("inventory resource not found, has been deleted")
+			return ctrl.Result{}, err
+		}
+		logger.Error(err, "Error fetching CrunchyBridgeInstance object for reconciliation")
+		return ctrl.Result{}, err
+	}
+	bridgeapiClient, err := setupClient(r.Client, inventory, r.APIBaseURL, logger)
+	if err != nil {
+		statusErr := r.updateStatus(instanceObj, metav1.ConditionFalse, AuthenticationError, err.Error())
+		if statusErr != nil {
+			logger.Error(statusErr, "Error in updating CrunchyBridgeInstance status")
+			return ctrl.Result{Requeue: true}, statusErr
+		}
+		logger.Error(err, "No CrunchyBridge client configured")
+		return ctrl.Result{}, err
+	}
+	logger.Info("Crunchy Bridge Client Configured ")
+
+	if instanceObj.DeletionTimestamp != nil && !instanceObj.DeletionTimestamp.IsZero() {
+		// Cluster deletion request / process finalizer
+		if listContains(instanceObj.Finalizers, instanceFinalizer) {
+			if id := instanceObj.Status.InstanceID; id != "" {
+				logger.Info("deleting cluster", "id", id)
+				if err := bridgeapiClient.DeleteCluster(id); err != nil {
+					return ctrl.Result{}, err
+				}
+				logger.Info("cluster deleted", "id", id)
+			}
+
+			controllerutil.RemoveFinalizer(instanceObj, instanceFinalizer)
+			if err := r.Update(ctx, instanceObj); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		switch instanceObj.Status.Phase {
```

Comment:
If the user deletes the instance CR and subsequently (when the cluster is being deleted) the same CR is created, what's going to happen to the cluster? Will the deletion gets completed first and then the new cluster (with the same name) gets created?

Reply:
there is a finalizer on CR, so deletion performs first, then the user will be able to create the CR with the same name.

Reply:
What if the deletion operation in the finalizer fails or gets stuck? Will this cause the CR deletion to get stuck as well?

Reply:
that depends on the case, in which scenario cr get stuck, at the moment it works as expected.

Reply:
if bridgeapiClient.DeleteCluster() call (line #110) fails, the deletion may get stuck.

Reply:
so this expected behavior, if DeleteCluster fails then reconciling again, and it should not delete the CR until DeleteCluster return success.